src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -191,6 +140,84 @@ private bool TryValidatePseudoHeaders()
 
+            return TryValidatePath(pathSegment);
+        }
+
+        private bool TryValidateMethod()
+        {
+            // :method
+            _methodText = RequestHeaders[HeaderNames.Method].ToString();
+            Method = HttpUtilities.GetKnownMethod(_methodText);
+
+            if (Method == HttpMethod.None)
+            {
+                ResetAndAbort(new ConnectionAbortedException(CoreStrings.FormatHttp2ErrorMethodInvalid(_methodText)), Http2ErrorCode.PROTOCOL_ERROR);
+                return false;
+            }
+
+            if (Method == HttpMethod.Custom)
+            {
+                if (HttpCharacters.IndexOfInvalidTokenChar(_methodText) >= 0)
+                {
+                    ResetAndAbort(new ConnectionAbortedException(CoreStrings.FormatHttp2ErrorMethodInvalid(_methodText)), Http2ErrorCode.PROTOCOL_ERROR);
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private bool TryValidateAuthorityAndHost(out string hostText)
+        {
+            // :authority (optional)
+            // Prefer this over Host
+
+            var authority = RequestHeaders[HeaderNames.Authority];
+            var host = HttpRequestHeaders.HeaderHost;
+            if (!StringValues.IsNullOrEmpty(authority))
+            {
+                // https://tools.ietf.org/html/rfc7540#section-8.1.2.3
+                // Clients that generate HTTP/2 requests directly SHOULD use the ":authority"
+                // pseudo - header field instead of the Host header field.
+                // An intermediary that converts an HTTP/2 request to HTTP/1.1 MUST
+                // create a Host header field if one is not present in a request by
+                // copying the value of the ":authority" pseudo - header field.
+
+                // We take this one step further, we don't want mismatched :authority
+                // and Host headers, replace Host if :authority is defined. The application
+                // will operate on the Host header.
+                HttpRequestHeaders.HeaderHost = authority;
+                host = authority;
+            }
+
+            // https://tools.ietf.org/html/rfc7230#section-5.4
+            // A server MUST respond with a 400 (Bad Request) status code to any
+            // HTTP/1.1 request message that lacks a Host header field and to any
+            // request message that contains more than one Host header field or a
+            // Host header field with an invalid field-value.
+            hostText = host.ToString();
+            if (host.Count > 1)
+            {
+                // RST replaces 400
+                ResetAndAbort(new ConnectionAbortedException(CoreStrings.FormatBadRequest_InvalidHostHeader_Detail(hostText)), Http2ErrorCode.PROTOCOL_ERROR);
+                return false;
+            }
+
+            try
+            {
+                HttpUtilities.ValidateHostHeader(hostText);
```

Comment:
I'd still like to see this change to an IsHostHeaderValid(hostText) call. We can keep ValidateHostHeader for HTTP/1.x and just have that also call IsHostHeaderValid.

Reply:
Updated